### DISCUSSION TITLE
"File exists" error when launching multiple MET/METplus tasks simultaneously

### DISCUSSION THREAD

**Discussion by gsketefian** [dtcenter/METplus#1733] [Statistical Computation] [answered]
Hi,

I am using the [rocoto](http://christopherwharrop.github.io/rocoto/) workflow manager to launch several MET/METplus vx tasks at (about) the same time.  Some of these tasks place their output in the same directory.  I think because of this, I sometimes get errors of this type:
```
ERROR: run_metplus  failed: [Errno 17] File exists: '/path/to/my/directories/expt_dirs/with_lag/2021050800/metprd/gen_ens_prod'
```
I think what’s happening is that two (or more) tasks notice that the output directory doesn’t exist, then start creating it.  One of them creates it first, so when another one tries to also create it, it finds that the directory already exists and generates this error (race condition?).

Do you think that hypothesis is correct?  If so, is there a way to get around this in METplus?

Thanks.

**Comment by JohnHalleyGotway** [comment]
I've asked @georgemccabe to take a look at this potential "race condition" when creating output directories. He should get back to you soon on this.

**Comment by georgemccabe** [comment, marked as answer]
Hi @gsketefian,

This issue came up recently during a telecon and a GitHub issue was created to address it:
https://github.com/dtcenter/METplus/issues/1657

Enhancements to the wrapper logic are needed to ensure that directory creation will work properly if writing to the same area concurrently for all instances. I adjusted the priority of this issue since this is the 2nd time that this has come up recently, so this fix should be made available in the next beta release, which is METplus-5.0.0.-beta3.

A temporary workaround for this specific issue would be to set different values for OUTPUT_BASE for each instance that is run concurrently. I know this is not ideal because you likely want all of the output in the same directory. Another option could be to create the output directories before running to ensure that the wrappers will not attempt to create them.

Please keep in mind that the METplus wrappers have not been thoroughly tested using workflow managers like rocoto, so there may be other issues that may need to be resolved. Please let us know if you encounter any other deficiencies so we can resolve them.

**Comment by gsketefian** [comment]
Ok, thanks @georgemccabe.

Any estimate on when METplus-5.0.0.-beta3 will be released?

For anyone else who's running into this issue in rocoto, I'm thinking a more automated workaround would be to add another task in the workflow (call it `make_output_dir`) that just creates this directory.  The actual vx tasks would then have a dependency on this task and would get launched only if `make_output_dir` has completed successfully.

**Reply by georgemccabe** [reply]
The beta 3 release is tentatively scheduled for 9/14/2022.

The workaround you suggested sounds like a reasonable solution for now.
